src/pybind/mgr/dashboard/CMakeLists.txt

```diff
@@ -104,3 +95,8 @@ add_custom_target(mgr-dashboard-frontend-build
   WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}/src/pybind/mgr/dashboard/frontend)
 add_dependencies(tests mgr-dashboard-frontend-build)
 endif(WITH_MGR_DASHBOARD_FRONTEND AND NOT CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|AARCH64|arm|ARM")
+
+if(WITH_TESTS)
+  include(AddCephTest)
+  add_tox_test(mgr-dashboard ${CMAKE_CURRENT_SOURCE_DIR})
```

Comment:
BTW, you may find my approach in https://github.com/ceph/ceph/pull/28696/files#diff-ef521ddc6371a1edba3225f01084ae53R68 (however, I don't know how to implement default values in cmake keyword function arguments).

Reply:
@epuertat , 

> We may just rename dashboard's py*-cov to just py, and leave covreport&lint&check apart (are we enforcing or plan to, some coverage threshold?).

although i am all for your proposal, but this involves some changes on the dashboard's `tox.ini` side. so it's not in the scope of my change. can we leave this to your PR?

i am adding a `TOX_ENVS` argument to `add_tox_test()` function. and let it default to `py27` and/or `py3` depending on `WITH_PYTHON2` and `WITH_PYTHON3`. hopefully this can make your work a little bit simpler. my only concern is that the change of `run_tox.sh` is not 100% backward compatible. as one needs to specify the `--tox-envs` with this change. @sebastian-philipp what do you think?

Reply:
I'm fine with it! Could you add a small usage documentation including an example then?

Reply:
@sebastian-philipp added the `--help` option and usage.

Reply:
Hm, what was the problem adding py3 or py27 depending on the cmake variables as a default for tox-env?

Reply:
i think it's a better design -- keep the cmake variables in cmake script and pass the just-enough information to the `run_tox.sh` script.

Reply:
It's just that tox envronments actually depends on cmake, as cmake decides which python versions are built in the pybinds.

Reply:
yeah, that's why i pass the tox envlist to run_tox.sh.

Reply:
@tchaikov I'm figuring out the best way with less code/run-length. What about this?

```cmake
# py27,py3 based on WITH_PYTHON* options
add_tox_test(mgr-dashboard)

add_tox_test(mgr-dashboard-lint TOX_ENVS lint)
add_tox_test(mgr-dashboard-check TOX_ENVS check)
```

Reply:
@epuertat i don't like this approach for couple reasons

1. `add_tox_test()` tries to ready the venv as part of `make tests`. using multiple `add_tox_test()` calls for a single `tox.ini` would introduce some unnecessarily complexity to this function.
2. `TOX_ENVS` option was designed to accept multiple parameters. i think you can just tweak `add_tox_test()` a little bit, like
```diff
diff --git a/cmake/modules/AddCephTest.cmake b/cmake/modules/AddCephTest.cmake
index bb5936cf8b..b21ae66016 100644
--- a/cmake/modules/AddCephTest.cmake
+++ b/cmake/modules/AddCephTest.cmake
@@ -69,7 +69,7 @@ function(add_tox_test name)
     list(APPEND tox_envs py3)
   endif()
   if(DEFINED TOXTEST_TOX_ENVS)
-    set(tox_envs ${TOXTEST_TOX_ENVS})
+    list(APPEND tox_envs ${TOXTEST_TOX_ENVS})
   endif()
   string(REPLACE ";" "," tox_envs "${tox_envs}")
   add_custom_command(
```
3. it's more like a question =) why do we need this flexibility of individual targets for different env?

Reply:
@tchaikov ah, great. I assumed you didn't want to allow additive toxenvs, but overwriting the defaults. I just wanted to avoid having to auto-guess the py27/py3 everytime some extra toxenv was required.

Reply:
yeah, i don't. but to avoid repeating this auto-guess logic everywhere. i think we can make a compromise by putting this cleverness in a single place.

Reply:
@sebastian-philipp if you want to have that feature. i will revert that part and bring the auto-guess code back into `run_tox.sh`. but personally, i'd 1) avoid repeating 2) have nicely encapsulated components/tools.

Reply:
@tchaikov maybe I'm being super orthodox here, but given dashboard's toxenvs `py{27,3}, lint`, and `check` have different requirements, would it make sense to create a new `tox.ini` (and `add_tox_test`) in `dashboard/ci` (where `check` core script is located), with its corresponding `requirements.txt` (which are different from dashboard`s)?

Just for adding context, that script basically checks that every Ceph-Dashboard front-end component embedding a Grafana dashboard matches the UUID for an existing Grafana Dashboard UUID as defined in their JSON file.

Reply:
sounds good if you think it's better to have a dedicated env for "check". but please note, adding a new `tox.ini` requires us to bootstrap yet another new venv just for creating the tox envs with existing machinery.